subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -0,0 +1,229 @@
+	size_t tx_shm_size;
+	size_t rx_shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+
+static void mbox_callback_process(struct k_work *item)
+{
+	struct backend_data_t *dev_data = CONTAINER_OF(item, struct backend_data_t, mbox_work);
+	uint8_t cb_buffer[CB_BUF_SIZE] __aligned(4);
+
+	atomic_t state = atomic_get(&dev_data->state);
+	int len = icmsg_buf_read(dev_data->rx_ib, cb_buffer, CB_BUF_SIZE);
+
+	__ASSERT_NO_MSG(len <= CB_BUF_SIZE);
+
+	if (len == -EAGAIN) {
+		(void)k_work_submit(&dev_data->mbox_work);
+		return;
+	} else if (len <= 0) {
+		return;
+	}
+
+	if (state == ICMSG_STATE_READY) {
+		if (!dev_data->ops || !dev_data->ops->received) {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		dev_data->ops->received(cb_buffer, len, NULL);
+
+		/* Reading with NULL buffer to know if there are data in the buffer to be read. */
+		len = icmsg_buf_read(dev_data->rx_ib, NULL, 0);
+		if (len > 0) {
+			(void)k_work_submit(&dev_data->mbox_work);
+		}
+	} else {
+		__ASSERT_NO_MSG(state == ICMSG_STATE_BUSY);
+		if (len != sizeof(magic) || memcmp(magic, cb_buffer, len)) {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		if (!dev_data->ops || !dev_data->ops->bound) {
+			__ASSERT_NO_MSG(false);
+			return;
+		}
+
+		dev_data->ops->bound(NULL);
+		atomic_set(&dev_data->state, ICMSG_STATE_READY);
+	}
+}
+
+static void mbox_callback(const struct device *instance, uint32_t channel,
+			  void *user_data, struct mbox_msg *msg_data)
+{
+	struct backend_data_t *dev_data = user_data;
+
+	(void)k_work_submit(&dev_data->mbox_work);
+}
+
+static int mbox_init(const struct device *instance)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	int err;
+
+	k_work_init(&dev_data->mbox_work, mbox_callback_process);
+
+	err = mbox_register_callback(&conf->mbox_rx, mbox_callback, dev_data);
+	if (err != 0) {
+		return err;
+	}
+
+	return mbox_set_enabled(&conf->mbox_rx, 1);
+}
+
+static int register_ept(const struct device *instance, void **token,
+			const struct ipc_ept_cfg *cfg)
+{
+	const struct backend_config_t *conf = instance->config;
+	struct backend_data_t *dev_data = instance->data;
+	struct ipc_ept *ep = (struct ipc_ept *)token;
+	int ret;
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_BUSY) {
+		/* Endpoint is still being initialized. */
+		return -EBUSY;
+	}
+
+	if (atomic_get(&dev_data->state) == ICMSG_STATE_READY) {
+		/* Already registered. This backend support single ep. */
+		return -EALREADY;
+	}
+
+	ep->instance = instance;
+	dev_data->ops = &cfg->cb;
+
+	atomic_set(&dev_data->state, ICMSG_STATE_BUSY);
+
+	ret = mbox_init(instance);
+	if (ret) {
+		return ret;
+	}
+
+	ret = icmsg_buf_write(dev_data->tx_ib, magic, sizeof(magic));
+	if (ret < sizeof(magic)) {
+		__ASSERT_NO_MSG(ret == sizeof(magic));
+		return ret;
+	}
+
+	ret = mbox_send(&conf->mbox_tx, NULL);
+	if (ret) {
+		return ret;
+	}
+
+	(void)k_work_submit(&dev_data->mbox_work);
```

Comment:
is it necessary? is it to handle case when event came to mbox before it got initialized?

Reply:
maybe we could do that only if there is pending data in the buffer?

Reply:
For the first question - yes. I didn't want to assume if mbox event would be stored even it arrived before it was initialized.
For the second - theoretically yes, but i wanted to always schedule as first thing the work does is check if data are in the buffer. It is low cost as it happens only once, but i understand the justification - i will change that.